resource/kind.go

```diff
@@ -0,0 +1,431 @@
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	_ Kind = &UntypedKind{}
+	_ Kind = &TypedKind[*UntypedObject]{}
+)
+
+// KindReader describes any type capable of reading one or more Kind object bytes (in wire format) into a specific or generic Object2.
+// KindReaders may be specific to a particular group/version/kind, or generic across any or all of group, version, and kind.
+// See TypedKind and UntypedKind as examples of KindReader implementations.
+type KindReader interface {
+	// Read consumes the wire-format bytes contained in the io.Reader, and unmarshals them into an instance of the
+	// Kind as an Object2.
+	// It MAY return an error if the provided bytes are not of an expected group, version, and/or kind,
+	// and MUST return an error if the provided bytes are not of the proper shape to be unmarshaled as a kind, or cannot be unmarshaled for any other reason.
+	// TODO: should this _also_ consume a wire format for identifying the type of bytes on the wire, or always assume JSON?
```

Comment:
The problem here being that the code invoking the reader is typically supplied the reader in advance, because kind reading may be non-standard (not just straight marshal/unmarshal). So the client has bytes of some content encoding that it knows (json, yaml, etc.), and was given a `KindReader` to use on creation, rather than at request time (to make the actual API for a user simpler).

Reply:
If that's the case then the only option would be to pass the encoding as an argument, yes.